// pako/lib/zlib/deflate.js
var n;
var o = require("../utils/common");
var i = require("./trees");
var a = require("./adler32");
var s = require("./crc32");
var l = require("./messages");
var c = 0;
var u = 4;
var d = 0;
var p = -2;
var f = -1;
var h = 4;
var m = 2;
var g = 8;
var v = 9;
var y = 286;
var b = 30;
var w = 19;
var S = y * 2 + 1;
var x = 15;
var A = 3;
var _ = 258;
var C = _ + A + 1;
var E = 42;
var k = 113;
var O = 1;
var T = 2;
var P = 3;
var R = 4;
function I(e, t) {
  e.msg = l[t];
  return t;
}
function D(e) {
  return (e << 1) - (e > 4 ? 9 : 0);
}
function M(e) {
  for (var t = e.length; --t >= 0;) {
    e[t] = 0;
  }
}
function N(e) {
  var t = e.state;
  var r = t.pending;
  if (r > e.avail_out) {
    r = e.avail_out;
  }
  if (r !== 0) {
    o.arraySet(e.output, t.pending_buf, t.pending_out, r, e.next_out);
    e.next_out += r;
    t.pending_out += r;
    e.total_out += r;
    e.avail_out -= r;
    t.pending -= r;
    if (t.pending === 0) {
      t.pending_out = 0;
    }
  }
}
function L(e, t) {
  i._tr_flush_block(e, e.block_start >= 0 ? e.block_start : -1, e.strstart - e.block_start, t);
  e.block_start = e.strstart;
  N(e.strm);
}
function j(e, t) {
  e.pending_buf[e.pending++] = t;
}
function $(e, t) {
  e.pending_buf[e.pending++] = t >>> 8 & 255;
  e.pending_buf[e.pending++] = t & 255;
}
function F(e, t) {
  var r;
  var n;
  var o = e.max_chain_length;
  var i = e.strstart;
  var a = e.prev_length;
  var s = e.nice_match;
  var l = e.strstart > e.w_size - C ? e.strstart - (e.w_size - C) : 0;
  var c = e.window;
  var u = e.w_mask;
  var d = e.prev;
  var p = e.strstart + _;
  var f = c[i + a - 1];
  var h = c[i + a];
  if (e.prev_length >= e.good_match) {
    o >>= 2;
  }
  if (s > e.lookahead) {
    s = e.lookahead;
  }
  do {
    if (c[(r = t) + a] === h && c[r + a - 1] === f && c[r] === c[i] && c[++r] === c[i + 1]) {
      i += 2;
      r++;
      do {} while (c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && c[++i] === c[++r] && i < p);
      n = _ - (p - i);
      i = p - _;
      if (a < n) {
        e.match_start = t;
        if (s <= (a = n)) {
          break;
        }
        f = c[i + a - 1];
        h = c[i + a];
      }
    }
  } while ((t = d[t & u]) > l && --o != 0);
  if (a <= e.lookahead) {
    return a;
  } else {
    return e.lookahead;
  }
}
function B(e) {
  var t;
  var r;
  var n;
  var i;
  var l;
  var c;
  var u;
  var d;
  var p;
  var f;
  var h = e.w_size;
  do {
    i = e.window_size - e.lookahead - e.strstart;
    if (e.strstart >= h + (h - C)) {
      o.arraySet(e.window, e.window, h, h, 0);
      e.match_start -= h;
      e.strstart -= h;
      e.block_start -= h;
      t = r = e.hash_size;
      while (n = e.head[--t], e.head[t] = h <= n ? n - h : 0, --r);
      for (t = r = h; n = e.prev[--t], e.prev[t] = h <= n ? n - h : 0, --r;);
      i += h;
    }
    if (e.strm.avail_in === 0) {
      break;
    }
    c = e.strm;
    u = e.window;
    d = e.strstart + e.lookahead;
    f = undefined;
    if ((p = i) < (f = c.avail_in)) {
      f = p;
    }
    r = f === 0 ? 0 : (c.avail_in -= f, o.arraySet(u, c.input, c.next_in, f, d), c.state.wrap === 1 ? c.adler = a(c.adler, u, f, d) : c.state.wrap === 2 && (c.adler = s(c.adler, u, f, d)), c.next_in += f, c.total_in += f, f);
    e.lookahead += r;
    if (e.lookahead + e.insert >= A) {
      l = e.strstart - e.insert;
      e.ins_h = e.window[l];
      e.ins_h = (e.ins_h << e.hash_shift ^ e.window[l + 1]) & e.hash_mask;
      while (e.insert && (e.ins_h = (e.ins_h << e.hash_shift ^ e.window[l + A - 1]) & e.hash_mask, e.prev[l & e.w_mask] = e.head[e.ins_h], e.head[e.ins_h] = l, l++, e.insert--, !(e.lookahead + e.insert < A)));
    }
  } while (e.lookahead < C && e.strm.avail_in !== 0);
}
function z(e, t) {
  var r;
  var n;
  while (true) {
    if (e.lookahead < C) {
      B(e);
      if (e.lookahead < C && t === c) {
        return O;
      }
      if (e.lookahead === 0) {
        break;
      }
    }
    r = 0;
    if (e.lookahead >= A) {
      e.ins_h = (e.ins_h << e.hash_shift ^ e.window[e.strstart + A - 1]) & e.hash_mask;
      r = e.prev[e.strstart & e.w_mask] = e.head[e.ins_h];
      e.head[e.ins_h] = e.strstart;
    }
    if (r !== 0 && e.strstart - r <= e.w_size - C) {
      e.match_length = F(e, r);
    }
    if (e.match_length >= A) {
      n = i._tr_tally(e, e.strstart - e.match_start, e.match_length - A);
      e.lookahead -= e.match_length;
      if (e.match_length <= e.max_lazy_match && e.lookahead >= A) {
        for (e.match_length--; e.strstart++, e.ins_h = (e.ins_h << e.hash_shift ^ e.window[e.strstart + A - 1]) & e.hash_mask, r = e.prev[e.strstart & e.w_mask] = e.head[e.ins_h], e.head[e.ins_h] = e.strstart, --e.match_length != 0;);
        e.strstart++;
      } else {
        e.strstart += e.match_length;
        e.match_length = 0;
        e.ins_h = e.window[e.strstart];
        e.ins_h = (e.ins_h << e.hash_shift ^ e.window[e.strstart + 1]) & e.hash_mask;
      }
    } else {
      n = i._tr_tally(e, 0, e.window[e.strstart]);
      e.lookahead--;
      e.strstart++;
    }
    if (n && (L(e, false), e.strm.avail_out === 0)) {
      return O;
    }
  }
  e.insert = e.strstart < A - 1 ? e.strstart : A - 1;
  if (t === u) {
    L(e, true);
    if (e.strm.avail_out === 0) {
      return P;
    } else {
      return R;
    }
  } else if (e.last_lit && (L(e, false), e.strm.avail_out === 0)) {
    return O;
  } else {
    return T;
  }
}
function U(e, t) {
  var r;
  var n;
  var o;
  while (true) {
    if (e.lookahead < C) {
      B(e);
      if (e.lookahead < C && t === c) {
        return O;
      }
      if (e.lookahead === 0) {
        break;
      }
    }
    r = 0;
    if (e.lookahead >= A) {
      e.ins_h = (e.ins_h << e.hash_shift ^ e.window[e.strstart + A - 1]) & e.hash_mask;
      r = e.prev[e.strstart & e.w_mask] = e.head[e.ins_h];
      e.head[e.ins_h] = e.strstart;
    }
    e.prev_length = e.match_length;
    e.prev_match = e.match_start;
    e.match_length = A - 1;
    if (r !== 0 && e.prev_length < e.max_lazy_match && e.strstart - r <= e.w_size - C) {
      e.match_length = F(e, r);
      if (e.match_length <= 5 && (e.strategy === 1 || e.match_length === A && e.strstart - e.match_start > 4096)) {
        e.match_length = A - 1;
      }
    }
    if (e.prev_length >= A && e.match_length <= e.prev_length) {
      o = e.strstart + e.lookahead - A;
      n = i._tr_tally(e, e.strstart - 1 - e.prev_match, e.prev_length - A);
      e.lookahead -= e.prev_length - 1;
      e.prev_length -= 2;
      while (++e.strstart <= o && (e.ins_h = (e.ins_h << e.hash_shift ^ e.window[e.strstart + A - 1]) & e.hash_mask, r = e.prev[e.strstart & e.w_mask] = e.head[e.ins_h], e.head[e.ins_h] = e.strstart), --e.prev_length != 0);
      e.match_available = 0;
      e.match_length = A - 1;
      e.strstart++;
      if (n && (L(e, false), e.strm.avail_out === 0)) {
        return O;
      }
    } else if (e.match_available) {
      if (n = i._tr_tally(e, 0, e.window[e.strstart - 1])) {
        L(e, false);
      }
      e.strstart++;
      e.lookahead--;
      if (e.strm.avail_out === 0) {
        return O;
      }
    } else {
      e.match_available = 1;
      e.strstart++;
      e.lookahead--;
    }
  }
  if (e.match_available) {
    n = i._tr_tally(e, 0, e.window[e.strstart - 1]);
    e.match_available = 0;
  }
  e.insert = e.strstart < A - 1 ? e.strstart : A - 1;
  if (t === u) {
    L(e, true);
    if (e.strm.avail_out === 0) {
      return P;
    } else {
      return R;
    }
  } else if (e.last_lit && (L(e, false), e.strm.avail_out === 0)) {
    return O;
  } else {
    return T;
  }
}
function H(e, t, r, n, o) {
  this.good_length = e;
  this.max_lazy = t;
  this.nice_length = r;
  this.max_chain = n;
  this.func = o;
}
function V() {
  this.strm = null;
  this.status = 0;
  this.pending_buf = null;
  this.pending_buf_size = 0;
  this.pending_out = 0;
  this.pending = 0;
  this.wrap = 0;
  this.gzhead = null;
  this.gzindex = 0;
  this.method = g;
  this.last_flush = -1;
  this.w_size = 0;
  this.w_bits = 0;
  this.w_mask = 0;
  this.window = null;
  this.window_size = 0;
  this.prev = null;
  this.head = null;
  this.ins_h = 0;
  this.hash_size = 0;
  this.hash_bits = 0;
  this.hash_mask = 0;
  this.hash_shift = 0;
  this.block_start = 0;
  this.match_length = 0;
  this.prev_match = 0;
  this.match_available = 0;
  this.strstart = 0;
  this.match_start = 0;
  this.lookahead = 0;
  this.prev_length = 0;
  this.max_chain_length = 0;
  this.max_lazy_match = 0;
  this.level = 0;
  this.strategy = 0;
  this.good_match = 0;
  this.nice_match = 0;
  this.dyn_ltree = new o.Buf16(S * 2);
  this.dyn_dtree = new o.Buf16((b * 2 + 1) * 2);
  this.bl_tree = new o.Buf16((w * 2 + 1) * 2);
  M(this.dyn_ltree);
  M(this.dyn_dtree);
  M(this.bl_tree);
  this.l_desc = null;
  this.d_desc = null;
  this.bl_desc = null;
  this.bl_count = new o.Buf16(x + 1);
  this.heap = new o.Buf16(y * 2 + 1);
  M(this.heap);
  this.heap_len = 0;
  this.heap_max = 0;
  this.depth = new o.Buf16(y * 2 + 1);
  M(this.depth);
  this.l_buf = 0;
  this.lit_bufsize = 0;
  this.last_lit = 0;
  this.d_buf = 0;
  this.opt_len = 0;
  this.static_len = 0;
  this.matches = 0;
  this.insert = 0;
  this.bi_buf = 0;
  this.bi_valid = 0;
}
function W(e) {
  var t;
  if (e && e.state) {
    e.total_in = e.total_out = 0;
    e.data_type = m;
    (t = e.state).pending = 0;
    t.pending_out = 0;
    if (t.wrap < 0) {
      t.wrap = -t.wrap;
    }
    t.status = t.wrap ? E : k;
    e.adler = t.wrap === 2 ? 0 : 1;
    t.last_flush = c;
    i._tr_init(t);
    return d;
  } else {
    return I(e, p);
  }
}
function q(e) {
  var t = W(e);
  if (t === d) {
    (function (e) {
      e.window_size = e.w_size * 2;
      M(e.head);
      e.max_lazy_match = n[e.level].max_lazy;
      e.good_match = n[e.level].good_length;
      e.nice_match = n[e.level].nice_length;
      e.max_chain_length = n[e.level].max_chain;
      e.strstart = 0;
      e.block_start = 0;
      e.lookahead = 0;
      e.insert = 0;
      e.match_length = e.prev_length = A - 1;
      e.match_available = 0;
      e.ins_h = 0;
    })(e.state);
  }
  return t;
}
function G(e, t, r, n, i, a) {
  if (!e) {
    return p;
  }
  var s = 1;
  if (t === f) {
    t = 6;
  }
  if (n < 0) {
    s = 0;
    n = -n;
  } else if (n > 15) {
    s = 2;
    n -= 16;
  }
  if (i < 1 || v < i || r !== g || n < 8 || n > 15 || t < 0 || t > 9 || a < 0 || h < a) {
    return I(e, p);
  }
  if (n === 8) {
    n = 9;
  }
  var l = new V();
  (e.state = l).strm = e;
  l.wrap = s;
  l.gzhead = null;
  l.w_bits = n;
  l.w_size = 1 << l.w_bits;
  l.w_mask = l.w_size - 1;
  l.hash_bits = i + 7;
  l.hash_size = 1 << l.hash_bits;
  l.hash_mask = l.hash_size - 1;
  l.hash_shift = ~~((l.hash_bits + A - 1) / A);
  l.window = new o.Buf8(l.w_size * 2);
  l.head = new o.Buf16(l.hash_size);
  l.prev = new o.Buf16(l.w_size);
  l.lit_bufsize = 1 << i + 6;
  l.pending_buf_size = l.lit_bufsize * 4;
  l.pending_buf = new o.Buf8(l.pending_buf_size);
  l.d_buf = l.lit_bufsize * 1;
  l.l_buf = l.lit_bufsize * 3;
  l.level = t;
  l.strategy = a;
  l.method = r;
  return q(e);
}
n = [new H(0, 0, 0, 0, function (e, t) {
  var r = 65535;
  for (r > e.pending_buf_size - 5 && (r = e.pending_buf_size - 5);;) {
    if (e.lookahead <= 1) {
      B(e);
      if (e.lookahead === 0 && t === c) {
        return O;
      }
      if (e.lookahead === 0) {
        break;
      }
    }
    e.strstart += e.lookahead;
    e.lookahead = 0;
    var n = e.block_start + r;
    if ((e.strstart === 0 || e.strstart >= n) && (e.lookahead = e.strstart - n, e.strstart = n, L(e, false), e.strm.avail_out === 0)) {
      return O;
    }
    if (e.strstart - e.block_start >= e.w_size - C && (L(e, false), e.strm.avail_out === 0)) {
      return O;
    }
  }
  e.insert = 0;
  if (t === u) {
    L(e, true);
    if (e.strm.avail_out === 0) {
      return P;
    } else {
      return R;
    }
  } else {
    if (e.strstart > e.block_start) {
      L(e, false);
      e.strm.avail_out;
    }
    return O;
  }
}), new H(4, 4, 8, 4, z), new H(4, 5, 16, 8, z), new H(4, 6, 32, 32, z), new H(4, 4, 16, 16, U), new H(8, 16, 32, 32, U), new H(8, 16, 128, 128, U), new H(8, 32, 128, 256, U), new H(32, 128, 258, 1024, U), new H(32, 258, 258, 4096, U)];
exports.deflateInit = function (e, t) {
  return G(e, t, g, 15, 8, 0);
};
exports.deflateInit2 = G;
exports.deflateReset = q;
exports.deflateResetKeep = W;
exports.deflateSetHeader = function (e, t) {
  if (e && e.state) {
    if (e.state.wrap !== 2) {
      return p;
    } else {
      e.state.gzhead = t;
      return d;
    }
  } else {
    return p;
  }
};
exports.deflate = function (e, t) {
  var r;
  var o;
  var a;
  var l;
  if (!e || !e.state || t > 5 || t < 0) {
    if (e) {
      return I(e, p);
    } else {
      return p;
    }
  }
  o = e.state;
  if (!e.output || !e.input && e.avail_in !== 0 || o.status === 666 && t !== u) {
    return I(e, e.avail_out === 0 ? -5 : p);
  }
  o.strm = e;
  r = o.last_flush;
  o.last_flush = t;
  if (o.status === E) {
    if (o.wrap === 2) {
      e.adler = 0;
      j(o, 31);
      j(o, 139);
      j(o, 8);
      if (o.gzhead) {
        j(o, (o.gzhead.text ? 1 : 0) + (o.gzhead.hcrc ? 2 : 0) + (o.gzhead.extra ? 4 : 0) + (o.gzhead.name ? 8 : 0) + (o.gzhead.comment ? 16 : 0));
        j(o, o.gzhead.time & 255);
        j(o, o.gzhead.time >> 8 & 255);
        j(o, o.gzhead.time >> 16 & 255);
        j(o, o.gzhead.time >> 24 & 255);
        j(o, o.level === 9 ? 2 : o.strategy >= 2 || o.level < 2 ? 4 : 0);
        j(o, o.gzhead.os & 255);
        if (o.gzhead.extra && o.gzhead.extra.length) {
          j(o, o.gzhead.extra.length & 255);
          j(o, o.gzhead.extra.length >> 8 & 255);
        }
        if (o.gzhead.hcrc) {
          e.adler = s(e.adler, o.pending_buf, o.pending, 0);
        }
        o.gzindex = 0;
        o.status = 69;
      } else {
        j(o, 0);
        j(o, 0);
        j(o, 0);
        j(o, 0);
        j(o, 0);
        j(o, o.level === 9 ? 2 : o.strategy >= 2 || o.level < 2 ? 4 : 0);
        j(o, 3);
        o.status = k;
      }
    } else {
      var f = g + (o.w_bits - 8 << 4) << 8;
      f |= (o.strategy >= 2 || o.level < 2 ? 0 : o.level < 6 ? 1 : o.level === 6 ? 2 : 3) << 6;
      if (o.strstart !== 0) {
        f |= 32;
      }
      f += 31 - f % 31;
      o.status = k;
      $(o, f);
      if (o.strstart !== 0) {
        $(o, e.adler >>> 16);
        $(o, e.adler & 65535);
      }
      e.adler = 1;
    }
  }
  if (o.status === 69) {
    if (o.gzhead.extra) {
      for (a = o.pending; o.gzindex < (o.gzhead.extra.length & 65535) && (o.pending !== o.pending_buf_size || (o.gzhead.hcrc && o.pending > a && (e.adler = s(e.adler, o.pending_buf, o.pending - a, a)), N(e), a = o.pending, o.pending !== o.pending_buf_size));) {
        j(o, o.gzhead.extra[o.gzindex] & 255);
        o.gzindex++;
      }
      if (o.gzhead.hcrc && o.pending > a) {
        e.adler = s(e.adler, o.pending_buf, o.pending - a, a);
      }
      if (o.gzindex === o.gzhead.extra.length) {
        o.gzindex = 0;
        o.status = 73;
      }
    } else {
      o.status = 73;
    }
  }
  if (o.status === 73) {
    if (o.gzhead.name) {
      a = o.pending;
      do {
        if (o.pending === o.pending_buf_size && (o.gzhead.hcrc && o.pending > a && (e.adler = s(e.adler, o.pending_buf, o.pending - a, a)), N(e), a = o.pending, o.pending === o.pending_buf_size)) {
          l = 1;
          break;
        }
        l = o.gzindex < o.gzhead.name.length ? o.gzhead.name.charCodeAt(o.gzindex++) & 255 : 0;
        j(o, l);
      } while (l !== 0);
      if (o.gzhead.hcrc && o.pending > a) {
        e.adler = s(e.adler, o.pending_buf, o.pending - a, a);
      }
      if (l === 0) {
        o.gzindex = 0;
        o.status = 91;
      }
    } else {
      o.status = 91;
    }
  }
  if (o.status === 91) {
    if (o.gzhead.comment) {
      a = o.pending;
      do {
        if (o.pending === o.pending_buf_size && (o.gzhead.hcrc && o.pending > a && (e.adler = s(e.adler, o.pending_buf, o.pending - a, a)), N(e), a = o.pending, o.pending === o.pending_buf_size)) {
          l = 1;
          break;
        }
        l = o.gzindex < o.gzhead.comment.length ? o.gzhead.comment.charCodeAt(o.gzindex++) & 255 : 0;
        j(o, l);
      } while (l !== 0);
      if (o.gzhead.hcrc && o.pending > a) {
        e.adler = s(e.adler, o.pending_buf, o.pending - a, a);
      }
      if (l === 0) {
        o.status = 103;
      }
    } else {
      o.status = 103;
    }
  }
  if (o.status === 103) {
    if (o.gzhead.hcrc) {
      if (o.pending + 2 > o.pending_buf_size) {
        N(e);
      }
      if (o.pending + 2 <= o.pending_buf_size) {
        j(o, e.adler & 255);
        j(o, e.adler >> 8 & 255);
        e.adler = 0;
        o.status = k;
      }
    } else {
      o.status = k;
    }
  }
  if (o.pending !== 0) {
    N(e);
    if (e.avail_out === 0) {
      o.last_flush = -1;
      return d;
    }
  } else if (e.avail_in === 0 && D(t) <= D(r) && t !== u) {
    return I(e, -5);
  }
  if (o.status === 666 && e.avail_in !== 0) {
    return I(e, -5);
  }
  if (e.avail_in !== 0 || o.lookahead !== 0 || t !== c && o.status !== 666) {
    var h = o.strategy === 2 ? function (e, t) {
      var r;
      for (;;) {
        if (e.lookahead === 0 && (B(e), e.lookahead === 0)) {
          if (t === c) {
            return O;
          }
          break;
        }
        e.match_length = 0;
        r = i._tr_tally(e, 0, e.window[e.strstart]);
        e.lookahead--;
        e.strstart++;
        if (r && (L(e, false), e.strm.avail_out === 0)) {
          return O;
        }
      }
      e.insert = 0;
      if (t === u) {
        L(e, true);
        if (e.strm.avail_out === 0) {
          return P;
        } else {
          return R;
        }
      } else if (e.last_lit && (L(e, false), e.strm.avail_out === 0)) {
        return O;
      } else {
        return T;
      }
    }(o, t) : o.strategy === 3 ? function (e, t) {
      var r;
      var n;
      var o;
      var a;
      var s = e.window;
      while (true) {
        if (e.lookahead <= _) {
          B(e);
          if (e.lookahead <= _ && t === c) {
            return O;
          }
          if (e.lookahead === 0) {
            break;
          }
        }
        e.match_length = 0;
        if (e.lookahead >= A && e.strstart > 0 && (n = s[o = e.strstart - 1]) === s[++o] && n === s[++o] && n === s[++o]) {
          a = e.strstart + _;
          do {} while (n === s[++o] && n === s[++o] && n === s[++o] && n === s[++o] && n === s[++o] && n === s[++o] && n === s[++o] && n === s[++o] && o < a);
          e.match_length = _ - (a - o);
          if (e.match_length > e.lookahead) {
            e.match_length = e.lookahead;
          }
        }
        if (e.match_length >= A) {
          r = i._tr_tally(e, 1, e.match_length - A);
          e.lookahead -= e.match_length;
          e.strstart += e.match_length;
          e.match_length = 0;
        } else {
          r = i._tr_tally(e, 0, e.window[e.strstart]);
          e.lookahead--;
          e.strstart++;
        }
        if (r && (L(e, false), e.strm.avail_out === 0)) {
          return O;
        }
      }
      e.insert = 0;
      if (t === u) {
        L(e, true);
        if (e.strm.avail_out === 0) {
          return P;
        } else {
          return R;
        }
      } else if (e.last_lit && (L(e, false), e.strm.avail_out === 0)) {
        return O;
      } else {
        return T;
      }
    }(o, t) : n[o.level].func(o, t);
    if (h === P || h === R) {
      o.status = 666;
    }
    if (h === O || h === P) {
      if (e.avail_out === 0) {
        o.last_flush = -1;
      }
      return d;
    }
    if (h === T && (t === 1 ? i._tr_align(o) : t !== 5 && (i._tr_stored_block(o, 0, 0, false), t === 3 && (M(o.head), o.lookahead === 0 && (o.strstart = 0, o.block_start = 0, o.insert = 0))), N(e), e.avail_out === 0)) {
      o.last_flush = -1;
      return d;
    }
  }
  if (t !== u) {
    return d;
  } else if (o.wrap <= 0) {
    return 1;
  } else {
    if (o.wrap === 2) {
      j(o, e.adler & 255);
      j(o, e.adler >> 8 & 255);
      j(o, e.adler >> 16 & 255);
      j(o, e.adler >> 24 & 255);
      j(o, e.total_in & 255);
      j(o, e.total_in >> 8 & 255);
      j(o, e.total_in >> 16 & 255);
      j(o, e.total_in >> 24 & 255);
    } else {
      $(o, e.adler >>> 16);
      $(o, e.adler & 65535);
    }
    N(e);
    if (o.wrap > 0) {
      o.wrap = -o.wrap;
    }
    if (o.pending !== 0) {
      return d;
    } else {
      return 1;
    }
  }
};
exports.deflateEnd = function (e) {
  var t;
  if (e && e.state) {
    if ((t = e.state.status) !== E && t !== 69 && t !== 73 && t !== 91 && t !== 103 && t !== k && t !== 666) {
      return I(e, p);
    } else {
      e.state = null;
      if (t === k) {
        return I(e, -3);
      } else {
        return d;
      }
    }
  } else {
    return p;
  }
};
exports.deflateSetDictionary = function (e, t) {
  var r;
  var n;
  var i;
  var s;
  var l;
  var c;
  var u;
  var f;
  var h = t.length;
  if (!e || !e.state) {
    return p;
  }
  if ((s = (r = e.state).wrap) === 2 || s === 1 && r.status !== E || r.lookahead) {
    return p;
  }
  if (s === 1) {
    e.adler = a(e.adler, t, h, 0);
  }
  r.wrap = 0;
  if (h >= r.w_size) {
    if (s === 0) {
      M(r.head);
      r.strstart = 0;
      r.block_start = 0;
      r.insert = 0;
    }
    f = new o.Buf8(r.w_size);
    o.arraySet(f, t, h - r.w_size, r.w_size, 0);
    t = f;
    h = r.w_size;
  }
  l = e.avail_in;
  c = e.next_in;
  u = e.input;
  e.avail_in = h;
  e.next_in = 0;
  e.input = t;
  B(r);
  while (r.lookahead >= A) {
    n = r.strstart;
    i = r.lookahead - (A - 1);
    while (r.ins_h = (r.ins_h << r.hash_shift ^ r.window[n + A - 1]) & r.hash_mask, r.prev[n & r.w_mask] = r.head[r.ins_h], r.head[r.ins_h] = n, n++, --i);
    r.strstart = n;
    r.lookahead = A - 1;
    B(r);
  }
  r.strstart += r.lookahead;
  r.block_start = r.strstart;
  r.insert = r.lookahead;
  r.lookahead = 0;
  r.match_length = r.prev_length = A - 1;
  r.match_available = 0;
  e.next_in = c;
  e.input = u;
  e.avail_in = l;
  r.wrap = s;
  return d;
};
exports.deflateInfo = "pako deflate (from Nodeca project)";
