// pako/lib/zlib/inflate.js
var n = require("../utils/common");
var o = require("./adler32");
var i = require("./crc32");
var a = require("./inffast");
var s = require("./inftrees");
var l = 1;
var c = 2;
var u = 0;
var d = -2;
var p = 1;
var f = 852;
var h = 592;
function m(e) {
  return (e >>> 24 & 255) + (e >>> 8 & 65280) + ((e & 65280) << 8) + ((e & 255) << 24);
}
function g() {
  this.mode = 0;
  this.last = false;
  this.wrap = 0;
  this.havedict = false;
  this.flags = 0;
  this.dmax = 0;
  this.check = 0;
  this.total = 0;
  this.head = null;
  this.wbits = 0;
  this.wsize = 0;
  this.whave = 0;
  this.wnext = 0;
  this.window = null;
  this.hold = 0;
  this.bits = 0;
  this.length = 0;
  this.offset = 0;
  this.extra = 0;
  this.lencode = null;
  this.distcode = null;
  this.lenbits = 0;
  this.distbits = 0;
  this.ncode = 0;
  this.nlen = 0;
  this.ndist = 0;
  this.have = 0;
  this.next = null;
  this.lens = new n.Buf16(320);
  this.work = new n.Buf16(288);
  this.lendyn = null;
  this.distdyn = null;
  this.sane = 0;
  this.back = 0;
  this.was = 0;
}
function v(e) {
  var t;
  if (e && e.state) {
    t = e.state;
    e.total_in = e.total_out = t.total = 0;
    e.msg = "";
    if (t.wrap) {
      e.adler = t.wrap & 1;
    }
    t.mode = p;
    t.last = 0;
    t.havedict = 0;
    t.dmax = 32768;
    t.head = null;
    t.hold = 0;
    t.bits = 0;
    t.lencode = t.lendyn = new n.Buf32(f);
    t.distcode = t.distdyn = new n.Buf32(h);
    t.sane = 1;
    t.back = -1;
    return u;
  } else {
    return d;
  }
}
function y(e) {
  var t;
  if (e && e.state) {
    (t = e.state).wsize = 0;
    t.whave = 0;
    t.wnext = 0;
    return v(e);
  } else {
    return d;
  }
}
function b(e, t) {
  var r;
  var n;
  if (e && e.state) {
    n = e.state;
    if (t < 0) {
      r = 0;
      t = -t;
    } else {
      r = 1 + (t >> 4);
      if (t < 48) {
        t &= 15;
      }
    }
    if (t && (t < 8 || t > 15)) {
      return d;
    } else {
      if (n.window !== null && n.wbits !== t) {
        n.window = null;
      }
      n.wrap = r;
      n.wbits = t;
      return y(e);
    }
  } else {
    return d;
  }
}
function w(e, t) {
  var r;
  var n;
  if (e) {
    n = new g();
    (e.state = n).window = null;
    if ((r = b(e, t)) !== u) {
      e.state = null;
    }
    return r;
  } else {
    return d;
  }
}
var S;
var x;
var A = true;
function _(e) {
  if (A) {
    var t;
    S = new n.Buf32(512);
    x = new n.Buf32(32);
    t = 0;
    while (t < 144) {
      e.lens[t++] = 8;
    }
    while (t < 256) {
      e.lens[t++] = 9;
    }
    while (t < 280) {
      e.lens[t++] = 7;
    }
    while (t < 288) {
      e.lens[t++] = 8;
    }
    s(l, e.lens, 0, 288, S, 0, e.work, {
      bits: 9
    });
    t = 0;
    while (t < 32) {
      e.lens[t++] = 5;
    }
    s(c, e.lens, 0, 32, x, 0, e.work, {
      bits: 5
    });
    A = false;
  }
  e.lencode = S;
  e.lenbits = 9;
  e.distcode = x;
  e.distbits = 5;
}
function C(e, t, r, o) {
  var i;
  var a = e.state;
  if (a.window === null) {
    a.wsize = 1 << a.wbits;
    a.wnext = 0;
    a.whave = 0;
    a.window = new n.Buf8(a.wsize);
  }
  if (o >= a.wsize) {
    n.arraySet(a.window, t, r - a.wsize, a.wsize, 0);
    a.wnext = 0;
    a.whave = a.wsize;
  } else {
    if (o < (i = a.wsize - a.wnext)) {
      i = o;
    }
    n.arraySet(a.window, t, r - o, i, a.wnext);
    if (o -= i) {
      n.arraySet(a.window, t, r - o, o, 0);
      a.wnext = o;
      a.whave = a.wsize;
    } else {
      a.wnext += i;
      if (a.wnext === a.wsize) {
        a.wnext = 0;
      }
      if (a.whave < a.wsize) {
        a.whave += i;
      }
    }
  }
  return 0;
}
exports.inflateReset = y;
exports.inflateReset2 = b;
exports.inflateResetKeep = v;
exports.inflateInit = function (e) {
  return w(e, 15);
};
exports.inflateInit2 = w;
exports.inflate = function (e, t) {
  var r;
  var f;
  var h;
  var g;
  var v;
  var y;
  var b;
  var w;
  var S;
  var x;
  var A;
  var E;
  var k;
  var O;
  var T;
  var P;
  var R;
  var I;
  var D;
  var M;
  var N;
  var L;
  var j;
  var $;
  var F = 0;
  var B = new n.Buf8(4);
  var z = [16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15];
  if (!e || !e.state || !e.output || !e.input && e.avail_in !== 0) {
    return d;
  }
  if ((r = e.state).mode === 12) {
    r.mode = 13;
  }
  v = e.next_out;
  h = e.output;
  b = e.avail_out;
  g = e.next_in;
  f = e.input;
  y = e.avail_in;
  w = r.hold;
  S = r.bits;
  x = y;
  A = b;
  L = u;
  e: while (true) {
    switch (r.mode) {
      case p:
        if (r.wrap === 0) {
          r.mode = 13;
          break;
        }
        while (S < 16) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if (r.wrap & 2 && w === 35615) {
          B[r.check = 0] = w & 255;
          B[1] = w >>> 8 & 255;
          r.check = i(r.check, B, 2, 0);
          S = w = 0;
          r.mode = 2;
          break;
        }
        r.flags = 0;
        if (r.head) {
          r.head.done = false;
        }
        if (!(r.wrap & 1) || (((w & 255) << 8) + (w >> 8)) % 31) {
          e.msg = "incorrect header check";
          r.mode = 30;
          break;
        }
        if ((w & 15) != 8) {
          e.msg = "unknown compression method";
          r.mode = 30;
          break;
        }
        S -= 4;
        N = 8 + ((w >>>= 4) & 15);
        if (r.wbits === 0) {
          r.wbits = N;
        } else if (N > r.wbits) {
          e.msg = "invalid window size";
          r.mode = 30;
          break;
        }
        r.dmax = 1 << N;
        e.adler = r.check = 1;
        r.mode = w & 512 ? 10 : 12;
        S = w = 0;
        break;
      case 2:
        while (S < 16) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        r.flags = w;
        if ((r.flags & 255) != 8) {
          e.msg = "unknown compression method";
          r.mode = 30;
          break;
        }
        if (r.flags & 57344) {
          e.msg = "unknown header flags set";
          r.mode = 30;
          break;
        }
        if (r.head) {
          r.head.text = w >> 8 & 1;
        }
        if (r.flags & 512) {
          B[0] = w & 255;
          B[1] = w >>> 8 & 255;
          r.check = i(r.check, B, 2, 0);
        }
        S = w = 0;
        r.mode = 3;
      case 3:
        while (S < 32) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if (r.head) {
          r.head.time = w;
        }
        if (r.flags & 512) {
          B[0] = w & 255;
          B[1] = w >>> 8 & 255;
          B[2] = w >>> 16 & 255;
          B[3] = w >>> 24 & 255;
          r.check = i(r.check, B, 4, 0);
        }
        S = w = 0;
        r.mode = 4;
      case 4:
        while (S < 16) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if (r.head) {
          r.head.xflags = w & 255;
          r.head.os = w >> 8;
        }
        if (r.flags & 512) {
          B[0] = w & 255;
          B[1] = w >>> 8 & 255;
          r.check = i(r.check, B, 2, 0);
        }
        S = w = 0;
        r.mode = 5;
      case 5:
        if (r.flags & 1024) {
          while (S < 16) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          r.length = w;
          if (r.head) {
            r.head.extra_len = w;
          }
          if (r.flags & 512) {
            B[0] = w & 255;
            B[1] = w >>> 8 & 255;
            r.check = i(r.check, B, 2, 0);
          }
          S = w = 0;
        } else if (r.head) {
          r.head.extra = null;
        }
        r.mode = 6;
      case 6:
        if (r.flags & 1024 && (y < (E = r.length) && (E = y), E && (r.head && (N = r.head.extra_len - r.length, r.head.extra ||= new Array(r.head.extra_len), n.arraySet(r.head.extra, f, g, E, N)), r.flags & 512 && (r.check = i(r.check, f, E, g)), y -= E, g += E, r.length -= E), r.length)) {
          break e;
        }
        r.length = 0;
        r.mode = 7;
      case 7:
        if (r.flags & 2048) {
          if (y === 0) {
            break e;
          }
          for (E = 0; N = f[g + E++], r.head && N && r.length < 65536 && (r.head.name += String.fromCharCode(N)), N && E < y;);
          if (r.flags & 512) {
            r.check = i(r.check, f, E, g);
          }
          y -= E;
          g += E;
          if (N) {
            break e;
          }
        } else if (r.head) {
          r.head.name = null;
        }
        r.length = 0;
        r.mode = 8;
      case 8:
        if (r.flags & 4096) {
          if (y === 0) {
            break e;
          }
          for (E = 0; N = f[g + E++], r.head && N && r.length < 65536 && (r.head.comment += String.fromCharCode(N)), N && E < y;);
          if (r.flags & 512) {
            r.check = i(r.check, f, E, g);
          }
          y -= E;
          g += E;
          if (N) {
            break e;
          }
        } else if (r.head) {
          r.head.comment = null;
        }
        r.mode = 9;
      case 9:
        if (r.flags & 512) {
          while (S < 16) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          if (w !== (r.check & 65535)) {
            e.msg = "header crc mismatch";
            r.mode = 30;
            break;
          }
          S = w = 0;
        }
        if (r.head) {
          r.head.hcrc = r.flags >> 9 & 1;
          r.head.done = true;
        }
        e.adler = r.check = 0;
        r.mode = 12;
        break;
      case 10:
        while (S < 32) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        e.adler = r.check = m(w);
        S = w = 0;
        r.mode = 11;
      case 11:
        if (r.havedict === 0) {
          e.next_out = v;
          e.avail_out = b;
          e.next_in = g;
          e.avail_in = y;
          r.hold = w;
          r.bits = S;
          return 2;
        }
        e.adler = r.check = 1;
        r.mode = 12;
      case 12:
        if (t === 5 || t === 6) {
          break e;
        }
      case 13:
        if (r.last) {
          w >>>= S & 7;
          S -= S & 7;
          r.mode = 27;
          break;
        }
        while (S < 3) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        r.last = w & 1;
        S -= 1;
        switch ((w >>>= 1) & 3) {
          case 0:
            r.mode = 14;
            break;
          case 1:
            _(r);
            r.mode = 20;
            if (t !== 6) {
              break;
            }
            w >>>= 2;
            S -= 2;
            break e;
          case 2:
            r.mode = 17;
            break;
          case 3:
            e.msg = "invalid block type";
            r.mode = 30;
        }
        w >>>= 2;
        S -= 2;
        break;
      case 14:
        w >>>= S & 7;
        S -= S & 7;
        while (S < 32) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if ((w & 65535) != (w >>> 16 ^ 65535)) {
          e.msg = "invalid stored block lengths";
          r.mode = 30;
          break;
        }
        r.length = w & 65535;
        S = w = 0;
        r.mode = 15;
        if (t === 6) {
          break e;
        }
      case 15:
        r.mode = 16;
      case 16:
        if (E = r.length) {
          if (y < E) {
            E = y;
          }
          if (b < E) {
            E = b;
          }
          if (E === 0) {
            break e;
          }
          n.arraySet(h, f, g, E, v);
          y -= E;
          g += E;
          b -= E;
          v += E;
          r.length -= E;
          break;
        }
        r.mode = 12;
        break;
      case 17:
        while (S < 14) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        r.nlen = 257 + (w & 31);
        w >>>= 5;
        S -= 5;
        r.ndist = 1 + (w & 31);
        w >>>= 5;
        S -= 5;
        r.ncode = 4 + (w & 15);
        w >>>= 4;
        S -= 4;
        if (r.nlen > 286 || r.ndist > 30) {
          e.msg = "too many length or distance symbols";
          r.mode = 30;
          break;
        }
        r.have = 0;
        r.mode = 18;
      case 18:
        while (r.have < r.ncode) {
          while (S < 3) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          r.lens[z[r.have++]] = w & 7;
          w >>>= 3;
          S -= 3;
        }
        while (r.have < 19) {
          r.lens[z[r.have++]] = 0;
        }
        r.lencode = r.lendyn;
        r.lenbits = 7;
        j = {
          bits: r.lenbits
        };
        L = s(0, r.lens, 0, 19, r.lencode, 0, r.work, j);
        r.lenbits = j.bits;
        if (L) {
          e.msg = "invalid code lengths set";
          r.mode = 30;
          break;
        }
        r.have = 0;
        r.mode = 19;
      case 19:
        while (r.have < r.nlen + r.ndist) {
          while (P = (F = r.lencode[w & (1 << r.lenbits) - 1]) >>> 16 & 255, R = F & 65535, !((T = F >>> 24) <= S)) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          if (R < 16) {
            w >>>= T;
            S -= T;
            r.lens[r.have++] = R;
          } else {
            if (R === 16) {
              for ($ = T + 2; S < $;) {
                if (y === 0) {
                  break e;
                }
                y--;
                w += f[g++] << S;
                S += 8;
              }
              w >>>= T;
              S -= T;
              if (r.have === 0) {
                e.msg = "invalid bit length repeat";
                r.mode = 30;
                break;
              }
              N = r.lens[r.have - 1];
              E = 3 + (w & 3);
              w >>>= 2;
              S -= 2;
            } else if (R === 17) {
              for ($ = T + 3; S < $;) {
                if (y === 0) {
                  break e;
                }
                y--;
                w += f[g++] << S;
                S += 8;
              }
              S -= T;
              N = 0;
              E = 3 + ((w >>>= T) & 7);
              w >>>= 3;
              S -= 3;
            } else {
              for ($ = T + 7; S < $;) {
                if (y === 0) {
                  break e;
                }
                y--;
                w += f[g++] << S;
                S += 8;
              }
              S -= T;
              N = 0;
              E = 11 + ((w >>>= T) & 127);
              w >>>= 7;
              S -= 7;
            }
            if (r.have + E > r.nlen + r.ndist) {
              e.msg = "invalid bit length repeat";
              r.mode = 30;
              break;
            }
            while (E--) {
              r.lens[r.have++] = N;
            }
          }
        }
        if (r.mode === 30) {
          break;
        }
        if (r.lens[256] === 0) {
          e.msg = "invalid code -- missing end-of-block";
          r.mode = 30;
          break;
        }
        r.lenbits = 9;
        j = {
          bits: r.lenbits
        };
        L = s(l, r.lens, 0, r.nlen, r.lencode, 0, r.work, j);
        r.lenbits = j.bits;
        if (L) {
          e.msg = "invalid literal/lengths set";
          r.mode = 30;
          break;
        }
        r.distbits = 6;
        r.distcode = r.distdyn;
        j = {
          bits: r.distbits
        };
        L = s(c, r.lens, r.nlen, r.ndist, r.distcode, 0, r.work, j);
        r.distbits = j.bits;
        if (L) {
          e.msg = "invalid distances set";
          r.mode = 30;
          break;
        }
        r.mode = 20;
        if (t === 6) {
          break e;
        }
      case 20:
        r.mode = 21;
      case 21:
        if (y >= 6 && b >= 258) {
          e.next_out = v;
          e.avail_out = b;
          e.next_in = g;
          e.avail_in = y;
          r.hold = w;
          r.bits = S;
          a(e, A);
          v = e.next_out;
          h = e.output;
          b = e.avail_out;
          g = e.next_in;
          f = e.input;
          y = e.avail_in;
          w = r.hold;
          S = r.bits;
          if (r.mode === 12) {
            r.back = -1;
          }
          break;
        }
        for (r.back = 0; P = (F = r.lencode[w & (1 << r.lenbits) - 1]) >>> 16 & 255, R = F & 65535, !((T = F >>> 24) <= S);) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if (P && !(P & 240)) {
          I = T;
          D = P;
          M = R;
          while (P = (F = r.lencode[M + ((w & (1 << I + D) - 1) >> I)]) >>> 16 & 255, R = F & 65535, !(I + (T = F >>> 24) <= S)) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          w >>>= I;
          S -= I;
          r.back += I;
        }
        w >>>= T;
        S -= T;
        r.back += T;
        r.length = R;
        if (P === 0) {
          r.mode = 26;
          break;
        }
        if (P & 32) {
          r.back = -1;
          r.mode = 12;
          break;
        }
        if (P & 64) {
          e.msg = "invalid literal/length code";
          r.mode = 30;
          break;
        }
        r.extra = P & 15;
        r.mode = 22;
      case 22:
        if (r.extra) {
          for ($ = r.extra; S < $;) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          r.length += w & (1 << r.extra) - 1;
          w >>>= r.extra;
          S -= r.extra;
          r.back += r.extra;
        }
        r.was = r.length;
        r.mode = 23;
      case 23:
        while (P = (F = r.distcode[w & (1 << r.distbits) - 1]) >>> 16 & 255, R = F & 65535, !((T = F >>> 24) <= S)) {
          if (y === 0) {
            break e;
          }
          y--;
          w += f[g++] << S;
          S += 8;
        }
        if (!(P & 240)) {
          I = T;
          D = P;
          M = R;
          while (P = (F = r.distcode[M + ((w & (1 << I + D) - 1) >> I)]) >>> 16 & 255, R = F & 65535, !(I + (T = F >>> 24) <= S)) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          w >>>= I;
          S -= I;
          r.back += I;
        }
        w >>>= T;
        S -= T;
        r.back += T;
        if (P & 64) {
          e.msg = "invalid distance code";
          r.mode = 30;
          break;
        }
        r.offset = R;
        r.extra = P & 15;
        r.mode = 24;
      case 24:
        if (r.extra) {
          for ($ = r.extra; S < $;) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          r.offset += w & (1 << r.extra) - 1;
          w >>>= r.extra;
          S -= r.extra;
          r.back += r.extra;
        }
        if (r.offset > r.dmax) {
          e.msg = "invalid distance too far back";
          r.mode = 30;
          break;
        }
        r.mode = 25;
      case 25:
        if (b === 0) {
          break e;
        }
        E = A - b;
        if (r.offset > E) {
          if ((E = r.offset - E) > r.whave && r.sane) {
            e.msg = "invalid distance too far back";
            r.mode = 30;
            break;
          }
          k = E > r.wnext ? (E -= r.wnext, r.wsize - E) : r.wnext - E;
          if (E > r.length) {
            E = r.length;
          }
          O = r.window;
        } else {
          O = h;
          k = v - r.offset;
          E = r.length;
        }
        if (b < E) {
          E = b;
        }
        b -= E;
        r.length -= E;
        while (h[v++] = O[k++], --E);
        if (r.length === 0) {
          r.mode = 21;
        }
        break;
      case 26:
        if (b === 0) {
          break e;
        }
        h[v++] = r.length;
        b--;
        r.mode = 21;
        break;
      case 27:
        if (r.wrap) {
          while (S < 32) {
            if (y === 0) {
              break e;
            }
            y--;
            w |= f[g++] << S;
            S += 8;
          }
          A -= b;
          e.total_out += A;
          r.total += A;
          if (A) {
            e.adler = r.check = r.flags ? i(r.check, h, A, v - A) : o(r.check, h, A, v - A);
          }
          A = b;
          if ((r.flags ? w : m(w)) !== r.check) {
            e.msg = "incorrect data check";
            r.mode = 30;
            break;
          }
          S = w = 0;
        }
        r.mode = 28;
      case 28:
        if (r.wrap && r.flags) {
          while (S < 32) {
            if (y === 0) {
              break e;
            }
            y--;
            w += f[g++] << S;
            S += 8;
          }
          if (w !== (r.total & 4294967295)) {
            e.msg = "incorrect length check";
            r.mode = 30;
            break;
          }
          S = w = 0;
        }
        r.mode = 29;
      case 29:
        L = 1;
        break e;
      case 30:
        L = -3;
        break e;
      case 31:
        return -4;
      default:
        return d;
    }
  }
  e.next_out = v;
  e.avail_out = b;
  e.next_in = g;
  e.avail_in = y;
  r.hold = w;
  r.bits = S;
  if ((r.wsize || A !== e.avail_out && r.mode < 30 && (r.mode < 27 || t !== 4)) && C(e, e.output, e.next_out, A - e.avail_out)) {
    r.mode = 31;
    return -4;
  } else {
    x -= e.avail_in;
    A -= e.avail_out;
    e.total_in += x;
    e.total_out += A;
    r.total += A;
    if (r.wrap && A) {
      e.adler = r.check = r.flags ? i(r.check, h, A, e.next_out - A) : o(r.check, h, A, e.next_out - A);
    }
    e.data_type = r.bits + (r.last ? 64 : 0) + (r.mode === 12 ? 128 : 0) + (r.mode === 20 || r.mode === 15 ? 256 : 0);
    if ((x == 0 && A === 0 || t === 4) && L === u) {
      L = -5;
    }
    return L;
  }
};
exports.inflateEnd = function (e) {
  if (!e || !e.state) {
    return d;
  }
  var t = e.state;
  t.window &&= null;
  e.state = null;
  return u;
};
exports.inflateGetHeader = function (e, t) {
  var r;
  if (e && e.state && (r = e.state).wrap & 2) {
    (r.head = t).done = false;
    return u;
  } else {
    return d;
  }
};
exports.inflateSetDictionary = function (e, t) {
  var r;
  var n = t.length;
  if (e && e.state) {
    if ((r = e.state).wrap !== 0 && r.mode !== 11) {
      return d;
    } else if (r.mode === 11 && o(1, t, n, 0) !== r.check) {
      return -3;
    } else if (C(e, t, n, n)) {
      r.mode = 31;
      return -4;
    } else {
      r.havedict = 1;
      return u;
    }
  } else {
    return d;
  }
};
exports.inflateInfo = "pako inflate (from Nodeca project)";
