// pako/lib/zlib/trees.js
var n = require("../utils/common");
var o = 0;
var i = 1;
function a(e) {
  for (var t = e.length; --t >= 0;) {
    e[t] = 0;
  }
}
var s = 0;
var l = 29;
var c = 256;
var u = c + 1 + l;
var d = 30;
var p = 19;
var f = u * 2 + 1;
var h = 15;
var m = 16;
var g = 7;
var v = 256;
var y = 16;
var b = 17;
var w = 18;
var S = [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 0];
var x = [0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13];
var A = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 3, 7];
var _ = [16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15];
var C = new Array((u + 2) * 2);
a(C);
var E = new Array(d * 2);
a(E);
var k = new Array(512);
a(k);
var O = new Array(256);
a(O);
var T = new Array(l);
a(T);
var P;
var R;
var I;
var D = new Array(d);
function M(e, t, r, n, o) {
  this.static_tree = e;
  this.extra_bits = t;
  this.extra_base = r;
  this.elems = n;
  this.max_length = o;
  this.has_stree = e && e.length;
}
function N(e, t) {
  this.dyn_tree = e;
  this.max_code = 0;
  this.stat_desc = t;
}
function L(e) {
  if (e < 256) {
    return k[e];
  } else {
    return k[256 + (e >>> 7)];
  }
}
function j(e, t) {
  e.pending_buf[e.pending++] = t & 255;
  e.pending_buf[e.pending++] = t >>> 8 & 255;
}
function $(e, t, r) {
  if (e.bi_valid > m - r) {
    e.bi_buf |= t << e.bi_valid & 65535;
    j(e, e.bi_buf);
    e.bi_buf = t >> m - e.bi_valid;
    e.bi_valid += r - m;
  } else {
    e.bi_buf |= t << e.bi_valid & 65535;
    e.bi_valid += r;
  }
}
function F(e, t, r) {
  $(e, r[t * 2], r[t * 2 + 1]);
}
function B(e, t) {
  for (var r = 0; r |= e & 1, e >>>= 1, r <<= 1, --t > 0;);
  return r >>> 1;
}
function z(e, t, r) {
  var n;
  var o;
  var i = new Array(h + 1);
  var a = 0;
  for (n = 1; n <= h; n++) {
    i[n] = a = a + r[n - 1] << 1;
  }
  for (o = 0; o <= t; o++) {
    var s = e[o * 2 + 1];
    if (s !== 0) {
      e[o * 2] = B(i[s]++, s);
    }
  }
}
function U(e) {
  var t;
  for (t = 0; t < u; t++) {
    e.dyn_ltree[t * 2] = 0;
  }
  for (t = 0; t < d; t++) {
    e.dyn_dtree[t * 2] = 0;
  }
  for (t = 0; t < p; t++) {
    e.bl_tree[t * 2] = 0;
  }
  e.dyn_ltree[v * 2] = 1;
  e.opt_len = e.static_len = 0;
  e.last_lit = e.matches = 0;
}
function H(e) {
  if (e.bi_valid > 8) {
    j(e, e.bi_buf);
  } else if (e.bi_valid > 0) {
    e.pending_buf[e.pending++] = e.bi_buf;
  }
  e.bi_buf = 0;
  e.bi_valid = 0;
}
function V(e, t, r, n) {
  var o = t * 2;
  var i = r * 2;
  return e[o] < e[i] || e[o] === e[i] && n[t] <= n[r];
}
function W(e, t, r) {
  for (var n = e.heap[r], o = r << 1; o <= e.heap_len && (o < e.heap_len && V(t, e.heap[o + 1], e.heap[o], e.depth) && o++, !V(t, n, e.heap[o], e.depth));) {
    e.heap[r] = e.heap[o];
    r = o;
    o <<= 1;
  }
  e.heap[r] = n;
}
function q(e, t, r) {
  var n;
  var o;
  var i;
  var a;
  var s = 0;
  if (e.last_lit !== 0) {
    while (n = e.pending_buf[e.d_buf + s * 2] << 8 | e.pending_buf[e.d_buf + s * 2 + 1], o = e.pending_buf[e.l_buf + s], s++, n === 0 ? F(e, o, t) : (F(e, (i = O[o]) + c + 1, t), (a = S[i]) !== 0 && $(e, o -= T[i], a), F(e, i = L(--n), r), (a = x[i]) !== 0 && $(e, n -= D[i], a)), s < e.last_lit);
  }
  F(e, v, t);
}
function G(e, t) {
  var r;
  var n;
  var o;
  var i = t.dyn_tree;
  var a = t.stat_desc.static_tree;
  var s = t.stat_desc.has_stree;
  var l = t.stat_desc.elems;
  var c = -1;
  e.heap_len = 0;
  e.heap_max = f;
  r = 0;
  for (; r < l; r++) {
    if (i[r * 2] !== 0) {
      e.heap[++e.heap_len] = c = r;
      e.depth[r] = 0;
    } else {
      i[r * 2 + 1] = 0;
    }
  }
  while (e.heap_len < 2) {
    i[(o = e.heap[++e.heap_len] = c < 2 ? ++c : 0) * 2] = 1;
    e.depth[o] = 0;
    e.opt_len--;
    if (s) {
      e.static_len -= a[o * 2 + 1];
    }
  }
  t.max_code = c;
  r = e.heap_len >> 1;
  for (; r >= 1; r--) {
    W(e, i, r);
  }
  for (o = l; r = e.heap[1], e.heap[1] = e.heap[e.heap_len--], W(e, i, 1), n = e.heap[1], e.heap[--e.heap_max] = r, e.heap[--e.heap_max] = n, i[o * 2] = i[r * 2] + i[n * 2], e.depth[o] = (e.depth[r] >= e.depth[n] ? e.depth[r] : e.depth[n]) + 1, i[r * 2 + 1] = i[n * 2 + 1] = o, e.heap[1] = o++, W(e, i, 1), e.heap_len >= 2;);
  e.heap[--e.heap_max] = e.heap[1];
  (function (e, t) {
    var r;
    var n;
    var o;
    var i;
    var a;
    var s;
    var l = t.dyn_tree;
    var c = t.max_code;
    var u = t.stat_desc.static_tree;
    var d = t.stat_desc.has_stree;
    var p = t.stat_desc.extra_bits;
    var m = t.stat_desc.extra_base;
    var g = t.stat_desc.max_length;
    var v = 0;
    for (i = 0; i <= h; i++) {
      e.bl_count[i] = 0;
    }
    l[e.heap[e.heap_max] * 2 + 1] = 0;
    r = e.heap_max + 1;
    for (; r < f; r++) {
      if (g < (i = l[l[(n = e.heap[r]) * 2 + 1] * 2 + 1] + 1)) {
        i = g;
        v++;
      }
      l[n * 2 + 1] = i;
      if (!(c < n)) {
        e.bl_count[i]++;
        a = 0;
        if (m <= n) {
          a = p[n - m];
        }
        s = l[n * 2];
        e.opt_len += s * (i + a);
        if (d) {
          e.static_len += s * (u[n * 2 + 1] + a);
        }
      }
    }
    if (v !== 0) {
      do {
        for (i = g - 1; e.bl_count[i] === 0;) {
          i--;
        }
        e.bl_count[i]--;
        e.bl_count[i + 1] += 2;
        e.bl_count[g]--;
        v -= 2;
      } while (v > 0);
      for (i = g; i !== 0; i--) {
        for (n = e.bl_count[i]; n !== 0;) {
          if (!(c < (o = e.heap[--r]))) {
            if (l[o * 2 + 1] !== i) {
              e.opt_len += (i - l[o * 2 + 1]) * l[o * 2];
              l[o * 2 + 1] = i;
            }
            n--;
          }
        }
      }
    }
  })(e, t);
  z(i, c, e.bl_count);
}
function X(e, t, r) {
  var n;
  var o;
  var i = -1;
  var a = t[1];
  var s = 0;
  var l = 7;
  var c = 4;
  if (a === 0) {
    l = 138;
    c = 3;
  }
  t[(r + 1) * 2 + 1] = 65535;
  n = 0;
  for (; n <= r; n++) {
    o = a;
    a = t[(n + 1) * 2 + 1];
    if (!(++s < l) || o !== a) {
      if (s < c) {
        e.bl_tree[o * 2] += s;
      } else if (o !== 0) {
        if (o !== i) {
          e.bl_tree[o * 2]++;
        }
        e.bl_tree[y * 2]++;
      } else if (s <= 10) {
        e.bl_tree[b * 2]++;
      } else {
        e.bl_tree[w * 2]++;
      }
      i = o;
      c = (s = 0) === a ? (l = 138, 3) : o === a ? (l = 6, 3) : (l = 7, 4);
    }
  }
}
function K(e, t, r) {
  var n;
  var o;
  var i = -1;
  var a = t[1];
  var s = 0;
  var l = 7;
  var c = 4;
  if (a === 0) {
    l = 138;
    c = 3;
  }
  n = 0;
  for (; n <= r; n++) {
    o = a;
    a = t[(n + 1) * 2 + 1];
    if (!(++s < l) || o !== a) {
      if (s < c) {
        while (F(e, o, e.bl_tree), --s != 0);
      } else if (o !== 0) {
        if (o !== i) {
          F(e, o, e.bl_tree);
          s--;
        }
        F(e, y, e.bl_tree);
        $(e, s - 3, 2);
      } else if (s <= 10) {
        F(e, b, e.bl_tree);
        $(e, s - 3, 3);
      } else {
        F(e, w, e.bl_tree);
        $(e, s - 11, 7);
      }
      i = o;
      c = (s = 0) === a ? (l = 138, 3) : o === a ? (l = 6, 3) : (l = 7, 4);
    }
  }
}
a(D);
var Y = false;
function J(e, t, r, o) {
  $(e, (s << 1) + (o ? 1 : 0), 3);
  (function (e, t, r, o) {
    H(e);
    if (o) {
      j(e, r);
      j(e, ~r);
    }
    n.arraySet(e.pending_buf, e.window, t, r, e.pending);
    e.pending += r;
  })(e, t, r, true);
}
exports._tr_init = function (e) {
  if (!Y) {
    (function () {
      var e;
      var t;
      var r;
      var n;
      var o;
      var i = new Array(h + 1);
      for (n = r = 0; n < l - 1; n++) {
        T[n] = r;
        e = 0;
        for (; e < 1 << S[n]; e++) {
          O[r++] = n;
        }
      }
      O[r - 1] = n;
      n = o = 0;
      for (; n < 16; n++) {
        D[n] = o;
        e = 0;
        for (; e < 1 << x[n]; e++) {
          k[o++] = n;
        }
      }
      for (o >>= 7; n < d; n++) {
        D[n] = o << 7;
        e = 0;
        for (; e < 1 << x[n] - 7; e++) {
          k[256 + o++] = n;
        }
      }
      for (t = 0; t <= h; t++) {
        i[t] = 0;
      }
      for (e = 0; e <= 143;) {
        C[e * 2 + 1] = 8;
        e++;
        i[8]++;
      }
      while (e <= 255) {
        C[e * 2 + 1] = 9;
        e++;
        i[9]++;
      }
      while (e <= 279) {
        C[e * 2 + 1] = 7;
        e++;
        i[7]++;
      }
      while (e <= 287) {
        C[e * 2 + 1] = 8;
        e++;
        i[8]++;
      }
      z(C, u + 1, i);
      e = 0;
      for (; e < d; e++) {
        E[e * 2 + 1] = 5;
        E[e * 2] = B(e, 5);
      }
      P = new M(C, S, c + 1, u, h);
      R = new M(E, x, 0, d, h);
      I = new M(new Array(0), A, 0, p, g);
    })();
    Y = true;
  }
  e.l_desc = new N(e.dyn_ltree, P);
  e.d_desc = new N(e.dyn_dtree, R);
  e.bl_desc = new N(e.bl_tree, I);
  e.bi_buf = 0;
  e.bi_valid = 0;
  U(e);
};
exports._tr_stored_block = J;
exports._tr_flush_block = function (e, t, r, n) {
  var a;
  var s;
  var l = 0;
  if (e.level > 0) {
    if (e.strm.data_type === 2) {
      e.strm.data_type = function (e) {
        var t;
        var r = 4093624447;
        for (t = 0; t <= 31; t++, r >>>= 1) {
          if (r & 1 && e.dyn_ltree[t * 2] !== 0) {
            return o;
          }
        }
        if (e.dyn_ltree[18] !== 0 || e.dyn_ltree[20] !== 0 || e.dyn_ltree[26] !== 0) {
          return i;
        }
        for (t = 32; t < c; t++) {
          if (e.dyn_ltree[t * 2] !== 0) {
            return i;
          }
        }
        return o;
      }(e);
    }
    G(e, e.l_desc);
    G(e, e.d_desc);
    l = function (e) {
      var t;
      X(e, e.dyn_ltree, e.l_desc.max_code);
      X(e, e.dyn_dtree, e.d_desc.max_code);
      G(e, e.bl_desc);
      t = p - 1;
      for (; t >= 3 && e.bl_tree[_[t] * 2 + 1] === 0; t--);
      e.opt_len += (t + 1) * 3 + 5 + 5 + 4;
      return t;
    }(e);
    a = e.opt_len + 3 + 7 >>> 3;
    if ((s = e.static_len + 3 + 7 >>> 3) <= a) {
      a = s;
    }
  } else {
    a = s = r + 5;
  }
  if (r + 4 <= a && t !== -1) {
    J(e, t, r, n);
  } else if (e.strategy === 4 || s === a) {
    $(e, 2 + (n ? 1 : 0), 3);
    q(e, C, E);
  } else {
    $(e, 4 + (n ? 1 : 0), 3);
    (function (e, t, r, n) {
      var o;
      $(e, t - 257, 5);
      $(e, r - 1, 5);
      $(e, n - 4, 4);
      o = 0;
      for (; o < n; o++) {
        $(e, e.bl_tree[_[o] * 2 + 1], 3);
      }
      K(e, e.dyn_ltree, t - 1);
      K(e, e.dyn_dtree, r - 1);
    })(e, e.l_desc.max_code + 1, e.d_desc.max_code + 1, l + 1);
    q(e, e.dyn_ltree, e.dyn_dtree);
  }
  U(e);
  if (n) {
    H(e);
  }
};
exports._tr_tally = function (e, t, r) {
  e.pending_buf[e.d_buf + e.last_lit * 2] = t >>> 8 & 255;
  e.pending_buf[e.d_buf + e.last_lit * 2 + 1] = t & 255;
  e.pending_buf[e.l_buf + e.last_lit] = r & 255;
  e.last_lit++;
  if (t === 0) {
    e.dyn_ltree[r * 2]++;
  } else {
    e.matches++;
    t--;
    e.dyn_ltree[(O[r] + c + 1) * 2]++;
    e.dyn_dtree[L(t) * 2]++;
  }
  return e.last_lit === e.lit_bufsize - 1;
};
exports._tr_align = function (e) {
  $(e, 2, 3);
  F(e, v, C);
  (function (e) {
    if (e.bi_valid === 16) {
      j(e, e.bi_buf);
      e.bi_buf = 0;
      e.bi_valid = 0;
    } else if (e.bi_valid >= 8) {
      e.pending_buf[e.pending++] = e.bi_buf & 255;
      e.bi_buf >>= 8;
      e.bi_valid -= 8;
    }
  })(e);
};
