// pako/lib/utils/common.js
var n = typeof Uint8Array != "undefined" && typeof Uint16Array != "undefined" && typeof Int32Array != "undefined";
exports.assign = function (e) {
  for (var t = Array.prototype.slice.call(arguments, 1); t.length;) {
    var r = t.shift();
    if (r) {
      if (typeof r != "object") {
        throw new TypeError(r + "must be non-object");
      }
      for (var n in r) {
        if (r.hasOwnProperty(n)) {
          e[n] = r[n];
        }
      }
    }
  }
  return e;
};
exports.shrinkBuf = function (e, t) {
  if (e.length === t) {
    return e;
  } else if (e.subarray) {
    return e.subarray(0, t);
  } else {
    e.length = t;
    return e;
  }
};
var o = {
  arraySet: function (e, t, r, n, o) {
    if (t.subarray && e.subarray) {
      e.set(t.subarray(r, r + n), o);
    } else {
      for (var i = 0; i < n; i++) {
        e[o + i] = t[r + i];
      }
    }
  },
  flattenChunks: function (e) {
    var t;
    var r;
    var n;
    var o;
    var i;
    var a;
    t = n = 0;
    r = e.length;
    for (; t < r; t++) {
      n += e[t].length;
    }
    a = new Uint8Array(n);
    t = o = 0;
    r = e.length;
    for (; t < r; t++) {
      i = e[t];
      a.set(i, o);
      o += i.length;
    }
    return a;
  }
};
var i = {
  arraySet: function (e, t, r, n, o) {
    for (var i = 0; i < n; i++) {
      e[o + i] = t[r + i];
    }
  },
  flattenChunks: function (e) {
    return [].concat.apply([], e);
  }
};
exports.setTyped = function (e) {
  if (e) {
    exports.Buf8 = Uint8Array;
    exports.Buf16 = Uint16Array;
    exports.Buf32 = Int32Array;
    exports.assign(exports, o);
  } else {
    exports.Buf8 = Array;
    exports.Buf16 = Array;
    exports.Buf32 = Array;
    exports.assign(exports, i);
  }
};
exports.setTyped(n);
