// pako/lib/utils/strings.js
var n = require("./common");
var o = true;
var i = true;
try {
  String.fromCharCode.apply(null, [0]);
} catch (e) {
  o = false;
}
try {
  String.fromCharCode.apply(null, new Uint8Array(1));
} catch (e) {
  i = false;
}
var a = new n.Buf8(256);
for (var s = 0; s < 256; s++) {
  a[s] = s >= 252 ? 6 : s >= 248 ? 5 : s >= 240 ? 4 : s >= 224 ? 3 : s >= 192 ? 2 : 1;
}
function l(e, t) {
  if (t < 65537 && (e.subarray && i || !e.subarray && o)) {
    return String.fromCharCode.apply(null, n.shrinkBuf(e, t));
  }
  var r = "";
  for (var a = 0; a < t; a++) {
    r += String.fromCharCode(e[a]);
  }
  return r;
}
a[254] = a[254] = 1;
exports.string2buf = function (e) {
  var t;
  var r;
  var o;
  var i;
  var a;
  var s = e.length;
  var l = 0;
  for (i = 0; i < s; i++) {
    if (((r = e.charCodeAt(i)) & 64512) == 55296 && i + 1 < s && ((o = e.charCodeAt(i + 1)) & 64512) == 56320) {
      r = 65536 + (r - 55296 << 10) + (o - 56320);
      i++;
    }
    l += r < 128 ? 1 : r < 2048 ? 2 : r < 65536 ? 3 : 4;
  }
  t = new n.Buf8(l);
  i = a = 0;
  for (; a < l; i++) {
    if (((r = e.charCodeAt(i)) & 64512) == 55296 && i + 1 < s && ((o = e.charCodeAt(i + 1)) & 64512) == 56320) {
      r = 65536 + (r - 55296 << 10) + (o - 56320);
      i++;
    }
    if (r < 128) {
      t[a++] = r;
    } else {
      if (r < 2048) {
        t[a++] = r >>> 6 | 192;
      } else {
        if (r < 65536) {
          t[a++] = r >>> 12 | 224;
        } else {
          t[a++] = r >>> 18 | 240;
          t[a++] = r >>> 12 & 63 | 128;
        }
        t[a++] = r >>> 6 & 63 | 128;
      }
      t[a++] = r & 63 | 128;
    }
  }
  return t;
};
exports.buf2binstring = function (e) {
  return l(e, e.length);
};
exports.binstring2buf = function (e) {
  var t = new n.Buf8(e.length);
  for (var r = 0, o = t.length; r < o; r++) {
    t[r] = e.charCodeAt(r);
  }
  return t;
};
exports.buf2string = function (e, t) {
  var r;
  var n;
  var o;
  var i;
  var s = t || e.length;
  var c = new Array(s * 2);
  for (r = n = 0; r < s;) {
    if ((o = e[r++]) < 128) {
      c[n++] = o;
    } else if ((i = a[o]) > 4) {
      c[n++] = 65533;
      r += i - 1;
    } else {
      for (o &= i === 2 ? 31 : i === 3 ? 15 : 7; i > 1 && r < s;) {
        o = o << 6 | e[r++] & 63;
        i--;
      }
      if (i > 1) {
        c[n++] = 65533;
      } else if (o < 65536) {
        c[n++] = o;
      } else {
        o -= 65536;
        c[n++] = o >> 10 & 1023 | 55296;
        c[n++] = o & 1023 | 56320;
      }
    }
  }
  return l(c, n);
};
exports.utf8border = function (e, t) {
  var r;
  if ((t = t || e.length) > e.length) {
    t = e.length;
  }
  r = t - 1;
  while (r >= 0 && (e[r] & 192) == 128) {
    r--;
  }
  if (r < 0 || r === 0) {
    return t;
  } else if (r + a[e[r]] > t) {
    return r;
  } else {
    return t;
  }
};
