// lie/index.js
var n = require("immediate");
function o() {}
var i = {};
var a = ["REJECTED"];
var s = ["FULFILLED"];
var l = ["PENDING"];
function c(e) {
  if (typeof e != "function") {
    throw new TypeError("resolver must be a function");
  }
  this.state = l;
  this.queue = [];
  this.outcome = undefined;
  if (e !== o) {
    f(this, e);
  }
}
function u(e, t, r) {
  this.promise = e;
  if (typeof t == "function") {
    this.onFulfilled = t;
    this.callFulfilled = this.otherCallFulfilled;
  }
  if (typeof r == "function") {
    this.onRejected = r;
    this.callRejected = this.otherCallRejected;
  }
}
function d(e, t, r) {
  n(function () {
    var n;
    try {
      n = t(r);
    } catch (n) {
      return i.reject(e, n);
    }
    if (n === e) {
      i.reject(e, new TypeError("Cannot resolve promise with itself"));
    } else {
      i.resolve(e, n);
    }
  });
}
function p(e) {
  var t = e && e.then;
  if (e && (typeof e == "object" || typeof e == "function") && typeof t == "function") {
    return function () {
      t.apply(e, arguments);
    };
  }
}
function f(e, t) {
  var r = false;
  function n(t) {
    if (!r) {
      r = true;
      i.reject(e, t);
    }
  }
  function o(t) {
    if (!r) {
      r = true;
      i.resolve(e, t);
    }
  }
  var a = h(function () {
    t(o, n);
  });
  if (a.status === "error") {
    n(a.value);
  }
}
function h(e, t) {
  var r = {};
  try {
    r.value = e(t);
    r.status = "success";
  } catch (e) {
    r.status = "error";
    r.value = e;
  }
  return r;
}
(module.exports = c).prototype.finally = function (e) {
  if (typeof e != "function") {
    return this;
  }
  var t = this.constructor;
  return this.then(function (r) {
    return t.resolve(e()).then(function () {
      return r;
    });
  }, function (r) {
    return t.resolve(e()).then(function () {
      throw r;
    });
  });
};
c.prototype.catch = function (e) {
  return this.then(null, e);
};
c.prototype.then = function (e, t) {
  if (typeof e != "function" && this.state === s || typeof t != "function" && this.state === a) {
    return this;
  }
  var r = new this.constructor(o);
  if (this.state !== l) {
    d(r, this.state === s ? e : t, this.outcome);
  } else {
    this.queue.push(new u(r, e, t));
  }
  return r;
};
u.prototype.callFulfilled = function (e) {
  i.resolve(this.promise, e);
};
u.prototype.otherCallFulfilled = function (e) {
  d(this.promise, this.onFulfilled, e);
};
u.prototype.callRejected = function (e) {
  i.reject(this.promise, e);
};
u.prototype.otherCallRejected = function (e) {
  d(this.promise, this.onRejected, e);
};
i.resolve = function (e, t) {
  var r = h(p, t);
  if (r.status === "error") {
    return i.reject(e, r.value);
  }
  var n = r.value;
  if (n) {
    f(e, n);
  } else {
    e.state = s;
    e.outcome = t;
    for (var o = -1, a = e.queue.length; ++o < a;) {
      e.queue[o].callFulfilled(t);
    }
  }
  return e;
};
i.reject = function (e, t) {
  e.state = a;
  e.outcome = t;
  for (var r = -1, n = e.queue.length; ++r < n;) {
    e.queue[r].callRejected(t);
  }
  return e;
};
c.resolve = function (e) {
  if (e instanceof this) {
    return e;
  } else {
    return i.resolve(new this(o), e);
  }
};
c.reject = function (e) {
  var t = new this(o);
  return i.reject(t, e);
};
c.all = function (e) {
  var t = this;
  if (Object.prototype.toString.call(e) !== "[object Array]") {
    return this.reject(new TypeError("must be an array"));
  }
  var r = e.length;
  var n = false;
  if (!r) {
    return this.resolve([]);
  }
  var a = new Array(r);
  var s = 0;
  for (var l = -1, c = new this(o); ++l < r;) {
    u(e[l], l);
  }
  return c;
  function u(e, o) {
    t.resolve(e).then(function (e) {
      a[o] = e;
      if (++s === r && !n) {
        n = true;
        i.resolve(c, a);
      }
    }, function (e) {
      if (!n) {
        n = true;
        i.reject(c, e);
      }
    });
  }
};
c.race = function (e) {
  var t = this;
  if (Object.prototype.toString.call(e) !== "[object Array]") {
    return this.reject(new TypeError("must be an array"));
  }
  var r = e.length;
  var n = false;
  if (!r) {
    return this.resolve([]);
  }
  var a;
  for (var s = -1, l = new this(o); ++s < r;) {
    a = e[s];
    t.resolve(a).then(function (e) {
      if (!n) {
        n = true;
        i.resolve(l, e);
      }
    }, function (e) {
      if (!n) {
        n = true;
        i.reject(l, e);
      }
    });
  }
  return l;
};
