// pako/lib/deflate.js
var n = require("./zlib/deflate");
var o = require("./utils/common");
var i = require("./utils/strings");
var a = require("./zlib/messages");
var s = require("./zlib/zstream");
var l = Object.prototype.toString;
var c = 0;
var u = -1;
var d = 0;
var p = 8;
function f(e) {
  if (!(this instanceof f)) {
    return new f(e);
  }
  this.options = o.assign({
    level: u,
    method: p,
    chunkSize: 16384,
    windowBits: 15,
    memLevel: 8,
    strategy: d,
    to: ""
  }, e || {});
  var t = this.options;
  if (t.raw && t.windowBits > 0) {
    t.windowBits = -t.windowBits;
  } else if (t.gzip && t.windowBits > 0 && t.windowBits < 16) {
    t.windowBits += 16;
  }
  this.err = 0;
  this.msg = "";
  this.ended = false;
  this.chunks = [];
  this.strm = new s();
  this.strm.avail_out = 0;
  var r = n.deflateInit2(this.strm, t.level, t.method, t.windowBits, t.memLevel, t.strategy);
  if (r !== c) {
    throw new Error(a[r]);
  }
  if (t.header) {
    n.deflateSetHeader(this.strm, t.header);
  }
  if (t.dictionary) {
    var h;
    h = typeof t.dictionary == "string" ? i.string2buf(t.dictionary) : l.call(t.dictionary) === "[object ArrayBuffer]" ? new Uint8Array(t.dictionary) : t.dictionary;
    if ((r = n.deflateSetDictionary(this.strm, h)) !== c) {
      throw new Error(a[r]);
    }
    this._dict_set = true;
  }
}
function h(e, t) {
  var r = new f(t);
  r.push(e, true);
  if (r.err) {
    throw r.msg || a[r.err];
  }
  return r.result;
}
f.prototype.push = function (e, t) {
  var r;
  var a;
  var s = this.strm;
  var u = this.options.chunkSize;
  if (this.ended) {
    return false;
  }
  a = t === ~~t ? t : t === true ? 4 : 0;
  if (typeof e == "string") {
    s.input = i.string2buf(e);
  } else if (l.call(e) === "[object ArrayBuffer]") {
    s.input = new Uint8Array(e);
  } else {
    s.input = e;
  }
  s.next_in = 0;
  s.avail_in = s.input.length;
  do {
    if (s.avail_out === 0) {
      s.output = new o.Buf8(u);
      s.next_out = 0;
      s.avail_out = u;
    }
    if ((r = n.deflate(s, a)) !== 1 && r !== c) {
      this.onEnd(r);
      return !(this.ended = true);
    }
    if (s.avail_out === 0 || s.avail_in === 0 && (a === 4 || a === 2)) {
      if (this.options.to === "string") {
        this.onData(i.buf2binstring(o.shrinkBuf(s.output, s.next_out)));
      } else {
        this.onData(o.shrinkBuf(s.output, s.next_out));
      }
    }
  } while ((s.avail_in > 0 || s.avail_out === 0) && r !== 1);
  if (a === 4) {
    r = n.deflateEnd(this.strm);
    this.onEnd(r);
    this.ended = true;
    return r === c;
  } else {
    return a !== 2 || (this.onEnd(c), !(s.avail_out = 0));
  }
};
f.prototype.onData = function (e) {
  this.chunks.push(e);
};
f.prototype.onEnd = function (e) {
  if (e === c) {
    if (this.options.to === "string") {
      this.result = this.chunks.join("");
    } else {
      this.result = o.flattenChunks(this.chunks);
    }
  }
  this.chunks = [];
  this.err = e;
  this.msg = this.strm.msg;
};
exports.Deflate = f;
exports.deflate = h;
exports.deflateRaw = function (e, t) {
  (t = t || {}).raw = true;
  return h(e, t);
};
exports.gzip = function (e, t) {
  (t = t || {}).gzip = true;
  return h(e, t);
};
