// pako/lib/inflate.js
var n = require("./zlib/inflate");
var o = require("./utils/common");
var i = require("./utils/strings");
var a = require("./zlib/constants");
var s = require("./zlib/messages");
var l = require("./zlib/zstream");
var c = require("./zlib/gzheader");
var u = Object.prototype.toString;
function d(e) {
  if (!(this instanceof d)) {
    return new d(e);
  }
  this.options = o.assign({
    chunkSize: 16384,
    windowBits: 0,
    to: ""
  }, e || {});
  var t = this.options;
  if (t.raw && t.windowBits >= 0 && t.windowBits < 16) {
    t.windowBits = -t.windowBits;
    if (t.windowBits === 0) {
      t.windowBits = -15;
    }
  }
  if (!!(t.windowBits >= 0) && !!(t.windowBits < 16) && (!e || !e.windowBits)) {
    t.windowBits += 32;
  }
  if (t.windowBits > 15 && t.windowBits < 48 && !(t.windowBits & 15)) {
    t.windowBits |= 15;
  }
  this.err = 0;
  this.msg = "";
  this.ended = false;
  this.chunks = [];
  this.strm = new l();
  this.strm.avail_out = 0;
  var r = n.inflateInit2(this.strm, t.windowBits);
  if (r !== a.Z_OK) {
    throw new Error(s[r]);
  }
  this.header = new c();
  n.inflateGetHeader(this.strm, this.header);
}
function p(e, t) {
  var r = new d(t);
  r.push(e, true);
  if (r.err) {
    throw r.msg || s[r.err];
  }
  return r.result;
}
d.prototype.push = function (e, t) {
  var r;
  var s;
  var l;
  var c;
  var d;
  var p;
  var f = this.strm;
  var h = this.options.chunkSize;
  var m = this.options.dictionary;
  var g = false;
  if (this.ended) {
    return false;
  }
  s = t === ~~t ? t : t === true ? a.Z_FINISH : a.Z_NO_FLUSH;
  if (typeof e == "string") {
    f.input = i.binstring2buf(e);
  } else if (u.call(e) === "[object ArrayBuffer]") {
    f.input = new Uint8Array(e);
  } else {
    f.input = e;
  }
  f.next_in = 0;
  f.avail_in = f.input.length;
  do {
    if (f.avail_out === 0) {
      f.output = new o.Buf8(h);
      f.next_out = 0;
      f.avail_out = h;
    }
    if ((r = n.inflate(f, a.Z_NO_FLUSH)) === a.Z_NEED_DICT && m) {
      p = typeof m == "string" ? i.string2buf(m) : u.call(m) === "[object ArrayBuffer]" ? new Uint8Array(m) : m;
      r = n.inflateSetDictionary(this.strm, p);
    }
    if (r === a.Z_BUF_ERROR && g === true) {
      r = a.Z_OK;
      g = false;
    }
    if (r !== a.Z_STREAM_END && r !== a.Z_OK) {
      this.onEnd(r);
      return !(this.ended = true);
    }
    if (f.next_out) {
      if (f.avail_out === 0 || r === a.Z_STREAM_END || f.avail_in === 0 && (s === a.Z_FINISH || s === a.Z_SYNC_FLUSH)) {
        if (this.options.to === "string") {
          l = i.utf8border(f.output, f.next_out);
          c = f.next_out - l;
          d = i.buf2string(f.output, l);
          f.next_out = c;
          f.avail_out = h - c;
          if (c) {
            o.arraySet(f.output, f.output, l, c, 0);
          }
          this.onData(d);
        } else {
          this.onData(o.shrinkBuf(f.output, f.next_out));
        }
      }
    }
    if (f.avail_in === 0 && f.avail_out === 0) {
      g = true;
    }
  } while ((f.avail_in > 0 || f.avail_out === 0) && r !== a.Z_STREAM_END);
  if (r === a.Z_STREAM_END) {
    s = a.Z_FINISH;
  }
  if (s === a.Z_FINISH) {
    r = n.inflateEnd(this.strm);
    this.onEnd(r);
    this.ended = true;
    return r === a.Z_OK;
  } else {
    return s !== a.Z_SYNC_FLUSH || (this.onEnd(a.Z_OK), !(f.avail_out = 0));
  }
};
d.prototype.onData = function (e) {
  this.chunks.push(e);
};
d.prototype.onEnd = function (e) {
  if (e === a.Z_OK) {
    if (this.options.to === "string") {
      this.result = this.chunks.join("");
    } else {
      this.result = o.flattenChunks(this.chunks);
    }
  }
  this.chunks = [];
  this.err = e;
  this.msg = this.strm.msg;
};
exports.Inflate = d;
exports.inflate = p;
exports.inflateRaw = function (e, t) {
  (t = t || {}).raw = true;
  return p(e, t);
};
exports.ungzip = p;
